### PR TITLE
Adds migrations to restore currupted staking ledgers in Polkadot and Kusama

### PR DESCRIPTION
**Pull request by gpestana**
Note: for more details on the corrupted ledgers issue and recovery steps check https://hackmd.io/m_h9DRutSZaUqCwM9tqZ3g?view.

This PR adds a migration in Polkadot and Kusama runtimes to recover the current corrupted ledgers in Polkadot and Kusama. A migration consists of:

1. Call into `pallet_staking::Pallet::<T>::restore_ledger` for each of the "whitelisted" stashes as `Root` origin.
2. Performs a check that ensures the restored ledger's stake does not overflow the current stash's free balance. If that's the case, force unstake the ledger. This check is currently missing in polkadot-sdk/pallet-staking ([PR with patch here](https://github.com/paritytech/polkadot-sdk/pull/5066)).

The reason to restore the corrupted ledgers as migrations implemented in the fellowship runtimes is to speed up the whole process and avoid having to wait for 1. merge and releases of https://github.com/paritytech/polkadot-sdk/pull/5066 and 2. referenda to call into `Call::restore_ledger` for both Polkadot and Kusama.

Alternatively, we could add a new temporary pallet directly in the fellowship runtime which would expose an extrinsic to restore the ledgers and perform the extra missing check. See this [PR as an example](https://github.com/gpestana/fellowship-runtimes/pull/2).

### CODE REVIEW

**Comment by bkchr** [comment]
This could just be an on chain referendum that calls the functions?

**Comment by gpestana** [comment]
> This could just be an on chain referendum that calls the functions?

We only can tell if a ledger needs force unstake at the time of recovery. We need to do the restore, check and potentially unstake atomically. In the time between the referenda are open and executed, new stashes can get in the situation that need to be force unstaked.

**Comment by gpestana** [comment]
> @bkchr'ss comment still holds: the flaw you are pointing out, the ledgers changing between the time of proposal and enactment, also applies to the hardcoded list of ledgers here.

So there's two things at play here:

1. The hardcoded list of accounts are fixed and should not change (since we fixed the runtime code so that there's no new corrupted ledgers).
2. From the set of hardcoded (and corrupted ledgers), some *may* need to be unstake. At this point, we know that one account needs to be unstake in Polkadot, but it may change in the future (if the stash changes its free balance so that free_balance < restored stake of the ledger).

**Comment by bkchr** [comment]
> Then I would rephrase the asnwer to @bkchr as: The logic here is a bit more than just what is available in a `pallet-staking` Call, therefore it has be a migration, or a temp pallet etc.

Bkchr says that he didn't know that the calls need to happen atomically. Then it makes sense to have them as a migration.

**Comment by mrshiposha** [comment]
Hi @gpestana.

Yesterday, we tried to stake DOT via a multisig account using `batchAll([staking.bond, staking.nominate])`.
We got `staking.BadState`. I thought it was some issue with `providers`/`consumers`, but they seem OK on our account (providers = 1).

Do you think this issue is related to the one this PR fixes, or is it a separate one?

Note: our account isn't part of the list from this PR.

**Comment by gpestana** [comment]
@mrshiposha a `staking.BadState` error is related to this PR yes. Do you want to reach out to me on Matrix so we can check the state of your account together? @gpestana:parity.io

**Review by kianenigma — APPROVED** [review]
LGTM except adding the migration sturcts to the migration tuple + reporting on a dry-run thereof.

**Comment by gpestana** [comment]
Just for the record, the issue that @mrshiposha has raised is not related to corrupted ledgers. The BadState error was raised due to trying to bond with a unfunded account as stash.

related to: https://github.com/paritytech/polkadot-sdk/pull/5627/files

File: relay/kusama/src/lib.rs [hunk 1]
```diff
@@ -1814,12 +1816,134 @@ pub mod migrations {
...
+
+	pub struct Migrate<T>(sp_std::marker::PhantomData<T>);
+	impl<T: pallet_staking::Config> OnRuntimeUpgrade for Migrate<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let mut total_weight = Default::default();
+
+			for stash in CorruptedStashes::get().into_iter() {
+				let stash_account: T::AccountId =
+					T::AccountId::decode(&mut &Into::<[u8; 32]>::into(stash.clone())[..])
+						.expect("32 bytes fits, qed.");
```
Comment (Ank4n) [inline relay/kusama/src/lib.rs:1868]: another way to build `AccountId`
`AccountId::from_str("ESGsxFePah1qb96ooTU4QJMxMKUG7NZvgTig3eJxP9f3wLa").unwrap();`

File: relay/kusama/src/lib.rs [hunk 2]
```diff
@@ -1814,12 +1816,134 @@ pub mod migrations {
...
+
+				// check if restored ledger total is higher than the stash's free balance. If
+				// that's the case, force unstake the ledger.
+				let weight = if let Ok(ledger) = pallet_staking::Pallet::<T>::ledger(
+					StakingAccount::Stash(stash_account.clone()),
+				) {
+					// force unstake the ledger.
+					if ledger.total > T::Currency::free_balance(&stash_account) {
+						let slashing_spans = 10; // default slashing spans for migration.
+						let _ = pallet_staking::Pallet::<T>::force_unstake(
```
Comment (Ank4n) [inline relay/kusama/src/lib.rs:1899]: Would it be better to force unstake only `ledger.total - free_balance`? Or the rationale is these ledgers are the overwritten ones and should be removed?

File: relay/kusama/src/lib.rs [hunk 3]
```diff
@@ -1814,12 +1816,134 @@ pub mod migrations {
...
+						target: LOG_TARGET,
+						"migrations::corrupted_ledgers: ledger does not exist, unexpected."
+					);
+					<T::WeightInfo>::restore_ledger()
+				};
+
+				total_weight += weight;
+			}
+
+			log::info!(target: LOG_TARGET, "migrations::corrupted_ledgers: done");
```
Comment (Ank4n) [inline relay/kusama/src/lib.rs:1941]: nit: useful generally to print a summary such as `success: x, errored: y`

**Review by Ank4n — APPROVED** [review]

**Comment by Juanma0x** [comment]
What else is needed for this PR to be merged? Are more approvals required, or does it depend on other developments?

File: relay/kusama/src/lib.rs (hunk 2 above)
Comment (gpestana) [reply relay/kusama/src/lib.rs:1899]: The rationale is that the new ledger does not have enough free balance to cover the stake in the ledger after recovery. For simplicity, we unstake the ledger (not critical since there's not much to unstake).

File: relay/kusama/src/lib.rs (hunk 1 above)
Comment (ggwpez) [inline relay/kusama/src/lib.rs:1868]: No unwrap in runtime code pls. I will push a fix

File: relay/kusama/src/lib.rs [hunk 4]
```diff
@@ -1814,12 +1816,134 @@ pub mod migrations {
...
+					}
+				} else {
+					log::error!(
+						target: LOG_TARGET,
+						"migrations::corrupted_ledgers: ledger does not exist, unexpected."
+					);
+					<T::WeightInfo>::restore_ledger()
+				};
+
+				total_weight += weight;
```
Comment (ggwpez) [inline relay/kusama/src/lib.rs:1938]:
```suggestion
				total_weight.saturating_accrue(weight);
```

Sometimes we use stuff like `u64::MAX` as XCM weight or in different spots.

File: relay/kusama/src/lib.rs [hunk 5]
```diff
@@ -1814,12 +1816,134 @@ pub mod migrations {
...
+						});
+
+						log::info!(
+							target: LOG_TARGET,
+							"migrations::corrupted_ledgers: ledger of {:?} restored (with force unstake).",
+							stash_account,
+						);
+
+						<T::WeightInfo>::restore_ledger() +
+							<T::WeightInfo>::force_unstake(slashing_spans)
```
Comment (ggwpez) [inline relay/kusama/src/lib.rs:1920]: Saturating add please.

File: relay/kusama/src/lib.rs (hunk 3 above)
Comment (gpestana) [reply relay/kusama/src/lib.rs:1941]: added!

**Comment by gpestana** [comment]
On runtime upgrade runs with a recent snapshot:

**Polkadot**
```bash
[2024-10-02T14:36:21Z INFO  try_runtime_core::common::misc_logging] ------------------------------------------------------------------


[2024-10-02T14:36:21Z INFO  try_runtime_core::common::misc_logging] 🔬 Running TryRuntime_on_runtime_upgrade with checks: PreAndPost


[2024-10-02T14:36:21Z INFO  try_runtime_core::common::misc_logging] ------------------------------------------------------------------


[2024-10-02T14:36:41Z INFO  runtime::polkadot] try-runtime::on_runtime_upgrade polkadot.

(...)

[2024-10-02T14:36:41Z INFO  runtime::polkadot] migrations::corrupted_ledgers: ledger of 5e510306a89f40e5520ae46adcc7a4a1bbacf27c86c163b0691bbbd7b5ef9c10 (5ECNRY7q...) restored (with force unstake).
[2024-10-02T14:36:41Z INFO  runtime::polkadot] migrations::corrupted_ledgers: ledger of a6379e16c5dab15e384c71024e3c6667356a5487127c291e61eed3d8d6b335dd (5FpeKyF3...) restored.
[2024-10-02T14:36:41Z INFO  runtime::polkadot] migrations::corrupted_ledgers: ledger of 6c3e8acb9225c2a6d22539e2c268c8721b016be1558b4aad4bed220dfbf01fea (5EWdcCGJ...) restored.
[2024-10-02T14:36:41Z INFO  runtime::polkadot] migrations::corrupted_ledgers: ledger of 4458ad5f0c082da64610607beb9d3164a77f1ef7964b7871c1de182ea7213783 (5DcKTQ71...) restored.
[2024-10-02T14:36:41Z INFO  runtime::polkadot] migrations::corrupted_ledgers: done. success: 4, error: 0
```

**Kusama**
```bash
[2024-10-02T14:39:54Z INFO  try_runtime_core::common::misc_logging] ------------------------------------------------------------------


[2024-10-02T14:39:54Z INFO  try_runtime_core::common::misc_logging] 🔬 Running TryRuntime_on_runtime_upgrade with checks: PreAndPost


[2024-10-02T14:39:54Z INFO  try_runtime_core::common::misc_logging] ------------------------------------------------------------------


[2024-10-02T14:40:19Z INFO  staging_kusama_runtime] try-runtime::on_runtime_upgrade kusama.

(...)

migrations::corrupted_ledgers: ledger of 52559f2c7324385aade778eca4d7837c7492d92ee79b66d6b416373066869d2e (5DvfDdum...) restored.
[2024-10-02T14:40:19Z INFO  runtime::kusama] migrations::corrupted_ledgers: ledger of 31162f413661f3f5351169299728ab6139725696ac6f98db9665e8b76d73d300 (5DB4nzJ4...) restored.
[2024-10-02T14:40:19Z INFO  runtime::kusama] migrations::corrupted_ledgers: ledger of 3a8012a52ec2715e711b1811f87684fe6646fc97a276043da7e75cd6a6516d29 (5DPQgPws...) restored.
[2024-10-02T14:40:19Z INFO  runtime::kusama] migrations::corrupted_ledgers: done. success: 3, error: 0
```
